File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.js
class Queue {
    constructor() {
        this.items = [];
        this.front = 0;
        this.rear = 0;
    }

    push(item) {
        this.items.push(item);
        this.rear++;
    }

    pop() {
        if (this.isEmpty()) {
            return null;
        }
        return this.items[this.front++];
    }

    isEmpty() {
        return this.front === this.rear;
    }
}

function solution(maps) {
    const [x_length, y_length] = [maps[0].length, maps.length];
    const directions = [[0, -1], [0, 1], [1, 0], [-1, 0]];
    const queue = new Queue();
    const visited = new Set();

    queue.push([0, 0, 1]);
    visited.add('0,0');

    while (!queue.isEmpty()) {
        const [x, y, dist] = queue.pop();

        if (x === x_length - 1 && y === y_length - 1) {
            return dist;
        }

        for (const [dx, dy] of directions) {
            const newX = x + dx;
            const newY = y + dy;

            if (newX >= 0 && newY >= 0 && newX < x_length && newY < y_length && maps[newY][newX] === 1) {
                const newPos = `${newX},${newY}`;
                if (!visited.has(newPos)) {
                    visited.add(newPos);
                    queue.push([newX, newY, dist + 1]);
                }
            }
        }
    }

    return -1;
}

const test_case = [
    [[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 1], [0, 0, 0, 0, 1]],
    [[1, 0, 1, 1, 1], [1, 0, 1, 0, 1], [1, 0, 1, 0, 1], [1, 0, 1, 1, 1], [1, 1, 1, 0, 0], [0, 0, 0, 0, 1]]
];

for (const t_case of test_case) {
    console.log(solution(t_case));
}

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque

def solution(maps):
    width = len(maps[0])
    height = len(maps)

    dx = [1, -1, 0, 0]
    dy = [0, 0, 1, -1]

    queue = deque([(0, 0, 1)])
    visited = [[False] * width for _ in range(height)]
    visited[0][0] = True

    while queue:
        x, y, count = queue.popleft()

        if x == width - 1 and y == height - 1:
            return count

        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]

            if 0 <= nx < width and 0 <= ny < height and not visited[ny][nx] and maps[ny][nx] == 1:
                visited[ny][nx] = True
                queue.append((nx, ny, count + 1))

    return -1